core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java

```diff
@@ -40,11 +48,37 @@ public ApplyUserAgentStage(HttpClientDependencies dependencies) {
     public SdkHttpFullRequest.Builder execute(SdkHttpFullRequest.Builder request, RequestExecutionContext context)
             throws Exception {
         RequestClientOptions opts = context.requestConfig().getRequestClientOptions();
-        if (opts != null) {
-            return request.header(HEADER_USER_AGENT, RuntimeHttpUtils
-                    .getUserAgent(clientConfig, opts.getClientMarker(RequestClientOptions.Marker.USER_AGENT)));
-        } else {
-            return request.header(HEADER_USER_AGENT, RuntimeHttpUtils.getUserAgent(clientConfig, null));
+        String userAgent = opts != null
+                           ? getUserAgent(clientConfig, opts.getClientMarker(RequestClientOptions.Marker.USER_AGENT))
+                           : getUserAgent(clientConfig, null);
+
+        return request.header(HEADER_USER_AGENT, userAgent);
+    }
+
+    private String getUserAgent(ClientConfiguration config, final String userAgentMarker) {
+        ClientOverrideConfiguration overrideConfig = config.overrideConfiguration();
+        String userDefinedPrefix = overrideConfig.advancedOption(AdvancedClientOption.USER_AGENT_PREFIX);
+        String userDefinedSuffix = overrideConfig.advancedOption(AdvancedClientOption.USER_AGENT_SUFFIX);
+        String awsExecutionEnvironment = AwsSystemSetting.AWS_EXECUTION_ENV.getStringValue().orElse(null);
+
+        StringBuilder userAgent = new StringBuilder(software.amazon.awssdk.utils.StringUtils.trimToEmpty(userDefinedPrefix));
+
+        if (!UserAgentUtils.getUserAgent().equals(userDefinedPrefix)) {
```

Comment:
This check seems unnecessary.

Reply:
I copied this. I'll look into it.

Reply:
Why does it seem uneccessary? I don't understand the format of the user agent well enough to know why it's not needed.

Reply:
So if I understand correctly we are just checking that the user supplied prefix doesn't exactly match our internal user agent. That seems like an error on the customers part if they do that for whatever reason.